connection/GraknTransaction.java

```diff
@@ -0,0 +1,635 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package grakn.client.connection;
+
+import com.google.protobuf.ByteString;
+import grabl.tracing.client.GrablTracingThreadStatic;
+import grakn.client.Grakn.Database;
+import grakn.client.Grakn.QueryOptions;
+import grakn.client.Grakn.Session;
+import grakn.client.Grakn.Transaction;
+import grakn.client.answer.Answer;
+import grakn.client.answer.AnswerGroup;
+import grakn.client.answer.ConceptList;
+import grakn.client.answer.ConceptMap;
+import grakn.client.answer.ConceptSet;
+import grakn.client.answer.ConceptSetMeasure;
+import grakn.client.answer.Explanation;
+import grakn.client.answer.Numeric;
+import grakn.client.answer.Void;
+import grakn.client.common.exception.GraknClientException;
+import grakn.client.concept.Concepts;
+import grakn.client.concept.connection.GraknConcepts;
+import grakn.protocol.AnswerProto;
+import grakn.protocol.ConceptProto;
+import grakn.protocol.GraknGrpc;
+import grakn.protocol.OptionsProto;
+import grakn.protocol.TransactionProto;
+import graql.lang.query.GraqlCompute;
+import graql.lang.query.GraqlDefine;
+import graql.lang.query.GraqlDelete;
+import graql.lang.query.GraqlGet;
+import graql.lang.query.GraqlInsert;
+import graql.lang.query.GraqlQuery;
+import graql.lang.query.GraqlUndefine;
+import io.grpc.ManagedChannel;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import static grabl.tracing.client.GrablTracingThreadStatic.traceOnThread;
+import static grakn.client.Grakn.Transaction.Type.READ;
+import static grakn.client.Grakn.Transaction.Type.WRITE;
+import static grakn.client.common.RpcMessageWriter.tracingData;
+import static grakn.client.common.exception.ErrorMessage.Query.UNRECOGNISED_QUERY_OBJECT;
+import static grakn.client.concept.ConceptMessageWriter.concept;
+import static grakn.client.connection.ConnectionMessageWriter.batchSize;
+import static grakn.client.connection.ConnectionMessageWriter.options;
+
+public class GraknTransaction implements Transaction {
+
+    private final Session session;
+    private final Transaction.Type type;
+    private final Concepts concepts;
+    private final GraknTransceiver transceiver;
+
+    public static class Builder implements Transaction.Builder {
+
+        private final ManagedChannel channel;
+        private final Session session;
+        private final ByteString sessionId;
+        private final QueryOptions options;
+
+        Builder(final ManagedChannel channel, final Session session, final ByteString sessionId) {
```

Comment:
Session can be changed to `GraknSession`, and then the `ManagedChannel` can be retrieved directly from session.